src/invariant/InvariantManagerImpl.cpp

```diff
@@ -0,0 +1,91 @@
+// Copyright 2017 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "invariant/InvariantDoesNotHold.h"
+#include "invariant/InvariantManagerImpl.h"
+#include "invariant/CacheIsConsistentWithDatabase.h"
+#include "invariant/ChangedAccountsSubentriesCountIsValid.h"
+#include "invariant/TotalCoinsEqualsBalancesPlusFeePool.h"
+#include "ledger/LedgerDelta.h"
+#include "lib/util/format.h"
+#include "util/Logging.h"
+#include "xdrpp/printer.h"
+#include <memory>
+
+namespace stellar
+{
+
+std::unique_ptr<InvariantManager>
+InvariantManager::create(Application& app)
+{
+    return make_unique<InvariantManagerImpl>();
+}
+
+InvariantManagerImpl::InvariantManagerImpl()
+{
+}
+
+void
+InvariantManagerImpl::checkOnLedgerClose(
+        TxSetFramePtr const& txSet,
+        LedgerDelta const& delta)
+{
+    for (auto invariant : mEnabled)
+    {
+        auto result = invariant->checkOnLedgerClose(delta);
+        if (result.empty())
+        {
+            continue;
+        }
+
+        auto transactions = TransactionSet{};
+        txSet->toXDR(transactions);
+        auto message =
+            fmt::format(R"(invariant "{}" does not hold on ledger {}: {}{}{})",
+                        invariant->getName(), delta.getHeader().ledgerSeq, result,
+                        "\n", xdr::xdr_to_string(transactions));
+        CLOG(FATAL, "Invariant") << message;
+        throw InvariantDoesNotHold{message};
+    }
+}
+
+void
+InvariantManagerImpl::registerInvariantImpl(
+        std::shared_ptr<Invariant> invariant)
+{
+    auto name = invariant->getName();
+    auto iter = mInvariants.find(name);
+    if (iter == mInvariants.end())
+    {
+        mInvariants[name] = invariant;
+    }
+    else
+    {
+        throw std::runtime_error{
+            "Invariant " + invariant->getName() + " already registered"};
+    }
+}
+
+void
+InvariantManagerImpl::enableInvariant(std::string const& name)
+{
+    auto invariant = mInvariants.at(name);
```

Comment:
it would be nice to throw a better exception than `out_of_range` when the invariant is unknown